Firmware/meatpack.cpp

```diff
@@ -0,0 +1,298 @@
+/*
+* MeatPack G-Code Compression
+*
+* Algorithm & Implementation: Scott Mudge - mail@scottmudge.com
+* Date: Dec. 2020
+*/
+
+#include "MeatPack.h"
+
+#ifdef ENABLE_MEATPACK
+
+#include "language.h"
+#include "Marlin.h"
+
+//#define MP_DEBUG
+
+// Utility definitions
+#define MeatPack_CommandByte		0b11111111
+#define MeatPack_NextPackedFirst	0b00000001
+#define MeatPack_NextPackedSecond	0b00000010
+
+
+// Note:
+// I've tried both a switch/case method and a lookup table. The disassembly is exactly the same after compilation, byte-to-byte.
+// Thus, performance is identical.
+#define USE_LOOKUP_TABLE
+
+#ifdef USE_LOOKUP_TABLE
+// The 15 most-common characters used in G-code, ~90-95% of all g-code uses these characters
+// NOT storing this with PROGMEM, given how frequently this table will be accessed.
+uint8_t MeatPackLookupTbl[16] = {
+    '0',	// 0000
+    '1',	// 0001
+    '2',	// 0010
+    '3',	// 0011
+    '4',	// 0100
+    '5',	// 0101
+    '6',	// 0110
+    '7',	// 0111
+    '8',	// 1000
+    '9',	// 1001
+    '.',	// 1010
+    ' ',	// 1011
+    '\n',	// 1100
+    'G',	// 1101
+    'X',	// 1110
+    '\0' // never used, 0b1111 is used to indicate next 8-bits is a full character
+};
+#else
+inline uint8_t get_char(register uint8_t in) {
+    switch (in) {
+    case 0b0000:
+        return '0';
+        break;
+    case 0b0001:
+        return '1';
+        break;
+    case 0b0010:
+        return '2';
+        break;
+    case 0b0011:
+        return '3';
+        break;
+    case 0b0100:
+        return '4';
+        break;
+    case 0b0101:
+        return '5';
+        break;
+    case 0b0110:
+        return '6';
+        break;
+    case 0b0111:
+        return '7';
+        break;
+    case 0b1000:
+        return '8';
+        break;
+    case 0b1001:
+        return '9';
+        break;
+    case 0b1010:
+        return '.';
+        break;
+    case 0b1011:
+        return ' ';
+        break;
+    case 0b1100:
+        return '\n';
+        break;
+    case 0b1101:
+        return 'G';
+        break;
+    case 0b1110:
+        return 'X';
+        break;
+    }
+    return 0;
+}
+#endif
+
+
+// State variables
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+volatile uint8_t mp_active = 0;             // Defines if MeatPack is active
```

Comment:
Why are these vars volatile? Are you planning to have them accessed from an ISR?

Reply:
99% of my embedded C experience has been with Cortex processors, and I rarely ever used the volatile keyword. For the Arduino compiler, documentation noted that using "volatile" was a way to force the variables out of registers, which Arduino had noted behave unpredictably. Given the critical nature of command parsing, I didn't find "unpredictable behavior" tolerable.

If I misinterpreted the Arduino documentation, feel free to modify the variable declarations.

Reply:
Alright, after reading up more on this, the primary Arduino documentation is simply being overly vague for the sake of not confusing junior developers.

`volatile` is indeed treated as it normally is, so I've corrected the variable qualifiers. I've also removed the public interfaces to any methods which could have been potentially used in ISRs to modify internal variables. They weren't used anyway. Will commit shortly.